### DISCUSSION TITLE
Missing permission. Redirect to login?

### DISCUSSION THREAD

**Discussion by joho1968** [kimai/kimai#4771] [General]
What would be a "normal" reason for the modal popping up with **Missing permission. Redirect to login?**

This is with 2.0.14, and I can get it in all sorts of places. Just now I got when I clicked on "Time Tracking > Calendar". The calendar **is** displayed behind the modal. I find no errors in any log file.

If I click **Confirm** in the modal dialog box, nothing happens. Or rather, I'm re-directed to the previous page I was on.

The only thing I can see in the nginx access log is a 403 for an API call:

`"GET /api/timesheets?user=4&size=1000&full=true&begin=2024-04-01T00:00:00&end=2024-05-13T00:00:00 HTTP/2.0" 403`

Can't figure this one out 🤔

**Comment by kevinpapst** [comment]
Check your browser / developer console and also check the network tab.
If you find the call that returns a permission problem, click on it and check the response.

**Reply by joho1968** [reply]
Well, the call is highlighted above. The response is "login-required: 1", among other headers. This is behind HTTP-auth, do I need to configure Kimai differently to send/keep the HTTP-auth headers?

**Reply by joho1968** [reply]
Though, looking at the request, there does seem to be `"Authorization: Basic <token>"` being passed ... 🤔

**Reply by joho1968** [reply]
This nginx block seems to be what's causing the issue:

    # Default to index.php, etc.
    location / {
        auth_basic "KIMAI.TEST";                                                                                       
        auth_basic_user_file /var/www/kimai.test/_auth/kimai;
        try_files $uri /index.php$is_args$args;
    }

If I remove this, and re-start nginx (and PHP), the problem persists. There seems to be something "hanging on", I don't know what or how, but if I kill the browser completely, and then re-start it, everything is fine 😕

**Reply by kevinpapst** [reply]
Okay, interesting use-case. I think I never tried that.
Without testing I would assume that the javascript requests made by Kimai do not have the Basic Auth header, so the requests fails on webserver level. And usually if a permission denied header is returned, Kimai assumes that the login expired and offers you to redirect to the login.

**Reply by joho1968** [reply]
Yes, I would assume this is what is happening. We quite frequently put stuff behind a first HTTP-AUTH, which is only used to protect the app/service in a basic manner (obviously). It does put a lot less load on the server since PHP never has to get involved to handle that.

Not sure what a good rule would be for nginx (and Apache), but I'm guessing it has `/api` at its beginning :sunglasses:

**Reply by joho1968** [reply]
Alternatively, is there any way Kimai could pick up whatever HTTP-AUTH headers are present at login and keep them in its session data and push it forward when making an (internal) API request :thinking:

**Reply by kevinpapst** [reply]
HTTP-AUTH is open for brute force attacks, unless you integrate fail2ban...
Block all and only allow access based on IPs is very effective as well.

**Reply by kevinpapst** [reply]
>  pick up whatever HTTP-AUTH headers are present at login and keep them in its session data and push it forward when making an (internal) API request 🤔

I guess it should be fairly easy for someone with javascript knowledge.
But I personally will not add a support layer for that.

**Reply by joho1968** [reply]
I know HTTP-AUTH is open for brute force attacks, but coupled with fail2ban it does make for a pretty good first layer of protection for apps that you may not want exposed on the Internet. So that's how we do it.

What I meant was that the PHP code could pick up these headers (if they're set) when you log in to Kimai, and then simply pass them on whenever you make an ajax call to fetch stuff via the API.

**Reply by joho1968** [reply]
@kevinpapst Are all API calls dispatched from a central place in the code, or are there multiple places (javascript) where these are made?

**Reply by kevinpapst** [reply]
Had a quick look, most should happen in these two places:
https://github.com/kimai/kimai/blob/main/assets/js/plugins/KimaiFetch.js
https://github.com/kimai/kimai/blob/main/assets/js/plugins/KimaiAPI.js

**Reply by joho1968** [reply]
I wonder if it's possible to do this with Apache/Nginx configuration directives somehow ... :thinking:

**Comment by nils-weidinger** [comment]
Being able to log in, but having "Missing permission. Redirect to login?" popping up all the time (even though edits were applied and everything was generally working) happened to me, when I made the mistake to use a soft link to subfolder "public" in my vhost configuration. Reasonably, that caused kimai to mess up the session data or something alike.

Directory content:
```
kimai -> kimai-2.44.0
kimai-2.44.0
site -> kimai/public
```

I was trying to use "site" as target directory in vhost configuration. Changing it to "kimai/public" solved the issue for me.
